src/tests/stub_libsecurity.cc

```diff
@@ -31,12 +31,30 @@ std::ostream &Security::operator <<(std::ostream &os, const Security::EncryptorA
 Security::HandshakeParser::HandshakeParser(MessageSource) STUB
 bool Security::HandshakeParser::parseHello(const SBuf &) STUB_RETVAL(false)
 
+#include "security/Io.h"
+Security::IoResult Security::Accept(Comm::Connection &) STUB_RETVAL(IoResult(IoResult::ioError))
+Security::IoResult Security::Connect(Comm::Connection &) STUB_RETVAL(IoResult(IoResult::ioError))
+void Security::ForgetErrors() STUB
+
 #include "security/KeyData.h"
 namespace Security
 {
 void KeyData::loadFromFiles(const AnyP::PortCfg &, const char *) STUB
 }
 
+#include "security/ErrorDetail.h"
+Security::ErrorDetail::ErrorDetail(ErrorCode, const CertPointer &, const CertPointer &, const char *) STUB
+#if USE_OPENSSL
```

Comment:
The library specific #if cases should never be exposed externally by libsecurity.

This one should be able to be removed by OpenSSL code defining LibErrorCode properly for the generic API users.

Reply:
We have many `#if USE_OPENSS` and `#if USE_GNUTLS` in squid code outside libsecurity.
Moreover the stub_libsecurity.cc is a dummy implementation of libsecurity. Why not to have this `#if` here?

Reply:
The existing #if outside of libsecurity are primarily legacy code that we are still needing to make generic.

The problem with having this if in the stub is that it wraps the entire definition. Exposing the fact that the class is not abstracted cleanly. As I said a clean abstraction would define LibErrorCode for OpenSSL code and this ctor could use it.

Reply:
My opinion is that we should consider the Security::ErrorDetail constructors  as internal to security namespace.  They are used only internally to generated errors. 
My sense is that their definition required in stub_libsecurity just because there is reference inside client_side.cc code,  inside `#if USE_OPENSSL` block, which is must moved under security namespace. When this is done these ErrorDetail constructors could be removed from stub_libsecurity.cc.

If required I think we can hide these two constructors by just build two new classes, one for each TLS toolkit which has only one constructor. But I am not seeing any strong reason for this. The code outside security and Ssl namespaces only interested for ErrorDetail::brief() and ErrorDetail::verbose() virtual methods, this is the abstract interface.

> As I said a clean abstraction would define LibErrorCode for OpenSSL code and this ctor could use it.

This branch adds a clear definition of LibErrorCode for OpenSSL. But GnuTLS and OpenSSL are different toolkits with their own philosophy and we can not have always the absolute equivalent of each other.
In Security::ErrorDetail constructor we need to pas an io error. The io error in GnuTLS is expressed by a LibErrorCode, but OpenSSL has a different set of integer io error codes.

I must also note that the previous try inside Squid to define a common error  code which is used by both OpenSSL and GnuTLS was the ErrorCode, but it was buggy.

But If you have suggestions I will like to hear about.